### DISCUSSION TITLE
How to migrate from pdfjs-dist 3.11.174 to version 4 on a node.js project

### DISCUSSION THREAD

**Discussion by inventivejon** [mozilla/pdf.js#17989] [Q&A]
I hope someone can help me with this issue.

I am using pdfjs-dist version 3.11.174 without any problems in my node.js 20.11.0 typescript project ("target": "esnext", "module": "commonjs",  "moduleResolution": "node",)

Now I wanted to upgrade to version 4. But there are only mjs file now.
The problem is that even the documentation for node.js uses
>> import { getDocument } from 'pdfjs-dist/legacy/build/pdf.mjs';

But when I try to do it this way I get an error from typescript:
Error [ERR_REQUIRE_ESM]: require() of ES Module .../node_modules/pdfjs-dist/build/pdf.mjs not supported.
Instead change the require of .../node_modules/pdfjs-dist/build/pdf.mjs to a dynamic import() which is available in all CommonJS modules.

So then I tried that:
import('pdfjs-dist/legacy/build/pdf.mjs').then(pdfjsLib => {
    const { getDocument } = pdfjsLib;
    // use getDocument here
}).catch(error => {
    console.error('Error loading the module', error);
});

But even then I get nearly the same error message. What do I need to change? Is there somewhere a migration document that explains how to migrate from 3 to 4 for a node.js project?

Thanks in advance.

**Comment by inventivejon** [comment]
Thanks for moving to correct position

**Comment by inventivejon** [comment]
Can nobody help?

**Comment by nicolo-ribaudo** [comment]
You need to set `"module": "preserve"` in your TS config, so that it doesn't transform your ESM imports to `require()` calls. You probably then also need to add `"type": "module"` to your package.json.

**Comment by djwlx** [comment]
I also want to ask, do I have to use the old version if I want to use it in CommonJs

**Comment by djwlx** [comment]
If not possible, I would like to know the latest version that supports commonjs

**Comment by ghost** [comment]
I have the same error when trying to upgrade pdfjs. But my porject is a bit different. 

I'm using `"module": "commonjs"` in my TS config, with node 20. I cannot change module resolution or use `type: "module"` in my package.json because the migration cost is huge, I get a lot of issues when trying.

So I tried to use a dynamic async `import` function call, like the error suggests. it compiles and works well locally with ts-node. But when the code is built by TS and runs, I get an error at runtime: "Promise.withResolvers is not a function"
<img width="991" alt="Screenshot 2024-05-13 at 16 17 51" src="https://github.com/mozilla/pdf.js/assets/63644124/92d04f2f-179a-4789-af80-5724ea48d95d">

And that is not surprising since this `Promise.withResolvers` is [only available from node 22](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/withResolvers). And since this is a dynamic import, TS cannot transpile / polyfill it

I think pdfjs is the only package we use that does not export a commonjs build, and I can't find a way to make it work properly.

Any idea?

**Reply by nicolo-ribaudo** [reply]
For the `Promise.withResolvers` problem, you should use the legacy build as documented at https://github.com/mozilla/pdf.js/wiki/Frequently-Asked-Questions#which-browsersenvironments-are-supported

**Reply by ghost** [reply]
Ah yes, I had missed this. Thank you!

**Reply by snowfluke** [reply]
could you share your code please, still experiencing the issue

**Reply by matthew-sutterlin** [reply]
^^ can you please share your code on how you set this up using the legacy builds, I cant seem to get this to work either when updating from 3 to 4 due to vulnerabilities

**Reply by ghost** [reply]
Oh sorry I didn't notice the replies to this comment.

Here's what I have:
```
async function importEsmModule<T>(name: string): Promise<T> {
  // FIXME: see https://stackoverflow.com/questions/65265420/how-to-prevent-typescript-from-transpiling-dynamic-imports-into-require
  // eslint-disable-next-line no-eval
  const module = eval(`(async () => {return await import("${name}")})()`);
  return module as T;
}

async function getPdf(params) {
  const { getDocument } = await importEsmModule<any>('pdfjs-dist/legacy/build/pdf.mjs');
  return getDocument(params)
}
```

I completely forgot how I came up with it, it's ugly as hell, but it works. And I only call the lib once so it's enough.

**Comment by lfoppiano** [comment]
Thanks to a friend with knowledge in javascript jungle, I have finally a working version. My project is a streamlit application but you can find the useful information in the repository, https://github.com/lfoppiano/streamlit-pdf-viewer/tree/enable-text-content-2 in particular the important changes: https://github.com/lfoppiano/streamlit-pdf-viewer/commit/5f09846fbc5b4cd71ab910f98b343d681e5808ed

**Comment by MrBOSS2142** [comment]
I have the same problem. I was using 3.11.174 version, but it have audit problem, so I need patched 4.x.x version. Does anyone have a solution?

**Reply by lfoppiano** [reply]
I'm not sure how much it could help, but we did it somehow, the project might be different, however. 

https://github.com/lfoppiano/streamlit-pdf-viewer/pull/52

**Reply by lfoppiano** [reply]
I see now that I provided a better link to the solution here: https://github.com/mozilla/pdf.js/discussions/17989#discussioncomment-9515932

**Comment by snowfluke** [comment]
Solved. For those who still getting the error, here's my solution:

tsconfig:
```ts
{
  "compilerOptions": {
    "module": "NodeNext",
    "esModuleInterop": true,
    "removeComments": true,
    "allowSyntheticDefaultImports": true,
    "emitDecoratorMetadata": true,
    "strict": true,
    "noImplicitOverride": true,
    "noPropertyAccessFromIndexSignature": true,
    "noImplicitReturns": true,
    "noFallthroughCasesInSwitch": true,
    "sourceMap": true,
    "declaration": false,
    "downlevelIteration": true,
    "experimentalDecorators": true,
    "moduleResolution": "NodeNext",
    "importHelpers": true,
    "useDefineForClassFields": false,
    "target": "ES2022",
    "outDir": "./dist",
    "allowJs": true,
    "baseUrl": "./",
    "incremental": true,
    "lib": ["dom"]
  },
  "include": ["**/*.ts"],
  "exclude": ["src/**/*.spec.ts", "./node_modules", "./dist"]
}
```

my import:

```ts
// import some type
import type { PDFDocumentProxy } from "pdfjs-dist";

// dynamic import
async function importPdfLib() {
  const pdfjsLib = await import("pdfjs-dist/legacy/build/pdf.mjs");
  return pdfjsLib;
}
```

**Reply by Erivess** [reply]
@snowfluke Thank you! This fixed my problem.

**Comment by xiaoyazhengit** [comment]
我有同样的问题，项目是vue2,js应该怎么修改呢

**Comment by guilherme-francisco** [comment]
This fixed my problem: https://github.com/TypeStrong/ts-node/discussions/1290#discussion-3301697

**Comment by plocket** [comment]
Is there still a legacy build available for v4 without typescript and without `import`?

**Reply by robross0606** [reply]
I have the same question.  It looks like usage in CommonJS projects is no longer possible since version 3.  Is that true?  It isn't well documented.

**Comment by inventivejon** [comment]
I actually now jumped from v3 to v5 directly and used node v22 and it worked again. Somehow v4 with old node version seems to cause the problems

**Reply by AliLee0923** [reply]
I am using "pdfjs-dist" in next.js api route and got this error.

Error extracting text from test.pdf: ReferenceError: DOMMatrix is not defined
    at <unknown> (webpack://pdf.js/src/display/canvas.js:59:21)
  57 | // Only used in rescaleAndStroke. The goal is to avoid
  58 | // creating a new DOMMatrix object each time we need it.
> 59 | const SCALE_MATRIX = new DOMMatrix();
     |                     ^
  60 |
  61 | // Used to get some coordinates.
  62 | const XY = new Float32Array(2); {
  page: '/api/parse-document'
}

Do you have any thought for this?

**Reply by snowfluke** [reply]
@AliLee0923  try my package https://www.npmjs.com/package/ppu-pdf

**Reply by AliLee0923** [reply]
ok, thank you.
let me try.
